### PR TITLE
Gh 1225 1 3 0 hashing

### PR DESCRIPTION
**Pull request by sche**
Referes #1225, #1226

Hashing and signing.

### CODE REVIEW

File: Multisig/Logic/Models/EIP712Transformer.swift [hunk 1]
```diff
@@ -7,18 +7,26 @@
...
         guard let safe = transaction.safe else { preconditionFailure("safe is required for typedData") }
 
-        let safeDomain = EIP712Domain(
+        var safeDomain = EIP712Domain(
             verifyingContract: safe.description
         )
+        var eip712Description = EIP712Domain.eip712Description
+
+        // Reference: https://github.com/gnosis/gnosis-py/blob/772a7da3a281c21931b8e5f01508a74cab7ecfb8/gnosis/safe/safe_tx.py#L169
+        if Version(transaction.safeVersion!)! >= Version("1.3.0")! {
```
Comment (DmitryBespalov) [inline Multisig/Logic/Models/EIP712Transformer.swift]: when can the `Version(transaction.safeVersion!)! ` crash?

File: Multisig/Logic/Models/Safe+CoreDataProperties.swift [hunk 2]
```diff
@@ -20,4 +20,5 @@ extension Safe {
     @NSManaged public var name: String?
     @NSManaged public var selection: Selection?
     @NSManaged public var network: Network?
+    @NSManaged public var contractVersion: String?
```
Comment (DmitryBespalov) [inline Multisig/Logic/Models/Safe+CoreDataProperties.swift:23]: is this a breaking change for the database?

File: Multisig/Logic/Models/Safe.swift [hunk 3]
```diff
@@ -42,11 +43,19 @@ extension Safe: Identifiable {
...
+    enum DomainSeparatorTypeHash {
+        static let v1_1_1 = Data(ethHex: "0x035aff83d86937d35b32e04f0ddc6ff469290eef2f1b692d8a815c89404d4749")
+        static let v1_3_0 = Data(ethHex: "0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218")
+    }
 
-    static func domainData(for safe: AddressString) -> Data {
-        Safe.DefaultEIP712SafeAppDomainSeparatorTypeHash + safe.data32
+    static func domainData(for safe: AddressString, version: String, chainId: String) -> Data {
+        let v = Version(version)
+        if let v = v, v >= Version("1.3.0")! {
```
Comment (DmitryBespalov) [inline Multisig/Logic/Models/Safe.swift]: a bit confused why not merge lines 52 and 53 😄

File: Multisig/Logic/Models/Safe.swift [hunk 4]
```diff
@@ -201,10 +211,14 @@ extension Safe: Identifiable {
 
 extension Safe {
     func update(from info: SafeInfoRequest.ResponseType) {
+        DispatchQueue.main.async {
```
Comment (DmitryBespalov) [inline Multisig/Logic/Models/Safe.swift:213]: why is this async? I thought just saving the object after assigning `contractVersion` would work

File: Multisig/Logic/Models/Safe.swift [hunk 5]
```diff
@@ -42,11 +43,19 @@ extension Safe: Identifiable {
...
-    static let DefaultEIP712SafeAppDomainSeparatorTypeHash =
-        Data(ethHex: "0x035aff83d86937d35b32e04f0ddc6ff469290eef2f1b692d8a815c89404d4749")
+    enum DomainSeparatorTypeHash {
+        static let v1_1_1 = Data(ethHex: "0x035aff83d86937d35b32e04f0ddc6ff469290eef2f1b692d8a815c89404d4749")
+        static let v1_3_0 = Data(ethHex: "0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218")
+    }
 
-    static func domainData(for safe: AddressString) -> Data {
-        Safe.DefaultEIP712SafeAppDomainSeparatorTypeHash + safe.data32
+    static func domainData(for safe: AddressString, version: String, chainId: String) -> Data {
```
Comment (DmitryBespalov) [inline Multisig/Logic/Models/Safe.swift:51]: now I noticed we have these stringy types: version and chainId, I think it'd be better to have actual types Version/SemVersion and ChainID, but that's for another refactoring

**Review by DmitryBespalov — COMMENTED** [review]
some comments

File: Multisig/Logic/Models/EIP712Transformer.swift (hunk 1 above)
Comment (sche) [reply Multisig/Logic/Models/EIP712Transformer.swift]: will make it safer

File: Multisig/Logic/Models/Safe+CoreDataProperties.swift (hunk 2 above)
Comment (sche) [reply Multisig/Logic/Models/Safe+CoreDataProperties.swift:23]: this is an additional field with default empty string value

File: Multisig/Logic/Models/Safe.swift (hunk 3 above)
Comment (sche) [reply Multisig/Logic/Models/Safe.swift]: ok

File: Multisig/Logic/Models/Safe.swift (hunk 4 above)
Comment (sche) [reply Multisig/Logic/Models/Safe.swift:213]: Update might happen on background thread.

**Review by DmitryBespalov — APPROVED** [review]

**Review by MouazAlzahabi — APPROVED** [review]

**Pull request merged**